python/ray/_private/gcs_pubsub.py

```diff
@@ -48,31 +49,107 @@ def construct_error_message(job_id, error_type, message, timestamp):
     return data
 
 
+class _PublisherBase:
+    def _create_log_request(self, log_batch: dict):
+        job_id = log_batch["job"]
+        return gcs_service_pb2.GcsPublishRequest(pub_messages=[
+            pubsub_pb2.PubMessage(
+                channel_type=pubsub_pb2.RAY_LOG_CHANNEL,
+                key_id=job_id.encode() if job_id else None,
+                log_batch_message=LogBatch(
+                    ip=str(log_batch["ip"]),
+                    # Cast to support string pid like "gcs".
+                    pid=str(log_batch["pid"]),
+                    # Job ID as a hex string.
+                    job_id=job_id,
+                    is_error=bool(log_batch["is_err"]),
+                    lines=log_batch["lines"],
+                    actor_name=log_batch["actor_name"],
+                    task_name=log_batch["task_name"],
+                ))
+        ])
+
+
 class _SubscriberBase:
+    def __init__(self):
```

Comment:
Core worker already has a PubSub client. Why not use that one? 
Can reduce the number of connections to GCS.

Reply:
Python publisher and subscribers are necessary for autoscaler and dashboard, which cannot depend on the worker library. @iycheng and I have discussed reusing gcs channel in core worker for Python workers. Because of the additional work needed to implement this optimization, we decided to just use a common GCS channel in Python pubsub and KV clients. In future we can implement an optimization to communicate to GCS via raylet.